Make browsertest failures standout

dhuang612 opened the pull request:
### Description of the Change

resolves issue #792 created a bar that starts out as green when the tests start.
Then if one of the tests fails bar will turn red to indicate test failure.
From the previous attempt learned that code coverage is decreased.
Will work on creating a test for this to get this resolved.

### Alternate Designs

<!-- Explain what other alternates were considered and why the proposed version was selected -->

### Why should this be in core?

<!-- Explain why this functionality should be in mocha as opposed to its own package -->

### Benefits

Allows for less eye movement to notice tests that don't pass when run from google chrome. 

### Possible Drawbacks

there shouldn't be any side effects because the changes are gui level

changes made are an enhancement.

dhuang612 commented:
reduced the size of the bar and gave them curved edges 

success:
<img width="1425" alt="Screen Shot 2020-01-20 at 1 46 38 PM" src="https://user-images.githubusercontent.com/49162249/72750698-8cfb6500-3b8b-11ea-9a25-1864cf35d891.png">

failure:
<img width="1407" alt="Screen Shot 2020-01-20 at 1 50 11 PM" src="https://user-images.githubusercontent.com/49162249/72750827-e368a380-3b8b-11ea-9553-b0325bb6ace1.png">

Please let me know if anything else needs to be changed

pedro-qwkin commented:
Why isn't this merged?

outsideris commented on lib/reporters/html.js:
```diff
@@ -35,7 +35,7 @@ exports = module.exports = HTML;
  * Stats template.
  */
-
+var statsBar = '<h1 id="progress-bar"></h1>';
```
How about we use `<div>` or `<progress>` tag since it is not heading.

dhuang612 replied:
@outsideris changed from h1 to a div

outsideris reviewed (commented):
It works with the latest master even we're using rollup instead browserify.

dhuang612 commented:
I tried to write some tests under base.spec they complete successfully but don't add to the coverage. Since there are no tests right now for the html browser. 
Let me know what you want done next.

juergba commented:
I recently pinned #3071 in the issue tracker and our exApplitools browser testing crack @giltayar commented: https://github.com/mochajs/mocha/issues/3071#issuecomment-863771039.

I'm aware of that we don't have any html reporter tests, yet. Wouldn't it be much easier to take the path Gil mentioned? A snapshot test with Selenium WebDriver? I'm attracted by Playwright though, I don't know exactly how it works, but it's nice 😎. We could forget about IE11 as we most probably are going to drop its support with the next major release v10.
We wouldn't need any mocking, bundling or Saucelab, just a HTML file with a failing test and compare the output against an existing snapshot.

juergba commented:
@dhuang612 I got a negative reply by the person I contacted, so the browser tests remains unsolved. He recommended to use playwright. I don't know yet how to proceed.

Your screenshots look nice, the colors look evtl. a bit strong. I haven't tested, yet.
`#mocha-stats.fail`: is there no other way like `extend`, more elegant than repeating the entire `#mocha-stats`? Is CSS so ugly?

dhuang612 commented:
I'll take a look at using extend.

dhuang612 commented:
@juergba 
I added in Sass as a dev package.
I then update the css file name to sass.
Finally, I switched the class I added to extend the original, and only updated the status bar colors themselves.

I tested the feature out using the example files to make sure they worked.

JoshuaKGoldberg commented on package.json:
```diff
@@ -144,6 +144,7 @@
     "rollup-plugin-node-globals": "^1.4.0",
     "rollup-plugin-node-polyfills": "^0.2.1",
     "rollup-plugin-visualizer": "^4.1.0",
+    "sass": "^1.39.2",
```
[Structure] `sass` is a pretty big dependency to use for exactly one `@extend`. I don't think it's justified here. Let's revert the change from `.css` to `.sass`, unless there's a strong reason for it?

mark-wiemer replied:
Removed in #5218 ;)

JoshuaKGoldberg commented on mocha.sass:
```diff
@@ -263,8 +263,18 @@ body {
+}
+
+#mocha-stats.fail {
+ @extend #mocha-stats;
```
[Style] I don't think this `@extend` is really needed. We can achieve the same result in raw CSS, such as with CSS variables or different class names. 

Thinking in terms of semantic class names, I think it'd make sense to have a class like `.fail`, `.pass`, etc. that each define their own colors.

Answering https://github.com/mochajs/mocha/pull/4148#issuecomment-915374638: no, CSS can overcome these issues. 🙂

JoshuaKGoldberg commented on lib/reporters/html.js:
```diff
@@ -111,7 +111,6 @@ function HTML(runner, options) {
       hideSuitesWithout('test fail');
     }
   });
-
```
[Style] Nit: unnecessary change?

JoshuaKGoldberg commented on mocha.sass:
Looking at the styles:

<table>
<thead>
<tr>
<th>Pass</th>
<th>Fail</th>
</tr>
</thead>
<tbody>
<td>
<img width="653" alt="image" src="https://github.com/mochajs/mocha/assets/3335181/09d5eefc-796c-42df-9e78-33c953593cf1">
</td>
<td>
<img width="653" alt="image" src="https://github.com/mochajs/mocha/assets/3335181/ced204f4-561d-4f73-9a9e-048aef41b8e6">
</td>
</tr>
</tbody>
</table>

A few things that make me not a fan yet:
* The new colors are hard to read on top of, to the point of not having [accessible color contrast](https://webaim.org/resources/contrastchecker/?fcolor=000000&bcolor=CC0000).
* The colors also don't match the existing aesthetic of the page. The page right now has very minimalist colors and a subtle design. This new green and red is pretty glaring. We should go with something that doesn't clash so much.
* It's a little inconsistent in how its laid out: more padding on the top vs. the bottom, and less horizontal padding than other UI elements.

I'm not sure what would be the "right" approach. A few things come to mind, but I haven't thought very deeply:

* Using much thinner border lines and/or much less saturated colors
    * Even switching to an [nprogress](https://github.com/rstacruz/nprogress)-style bar on the stats header or the page
* Instead of a background or border, using bold & colored text

(I'm not convinced any of that is better, just ideating)

JoshuaKGoldberg reviewed (changes requested):
I think we'll want to have some more discussion into what this UI should look like. Proposal: can we have that in the backing issue, #792? That way we can figure out what it should look like a bit more before jumping into the implementation.

JoshuaKGoldberg commented:
👋 ping @dhuang612, do you still have time for this?

JoshuaKGoldberg commented:
Closing out as it's been a while since PR activity. If anybody wants to take this over, please do - and post a co-author attribution if you take code from this PR. Cheers! 🤎

mark-wiemer commented:
@JoshuaKGoldberg I'll be coming to this one as a software engineer at Microsoft specializing in a11y :) please let me know if you have further comments before I start resolving the existing ones. I'll probably be asking for some help in the discord as well :) thanks again for picking up this project!
